File: app/src/main/java/com/amazon/tv/leanbacklauncher/ActiveItemsRowView.java
package com.amazon.tv.leanbacklauncher;

import android.content.Context;
import android.content.res.Resources;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewGroup.OnHierarchyChangeListener;

import androidx.leanback.widget.HorizontalGridView;

import com.amazon.tv.firetv.leanbacklauncher.apps.RowPreferences;
import com.amazon.tv.leanbacklauncher.animation.ParticipatesInScrollAnimation;
import com.amazon.tv.leanbacklauncher.animation.ViewDimmer;
import com.amazon.tv.leanbacklauncher.apps.AppsAdapter;
import com.amazon.tv.leanbacklauncher.capabilities.LauncherConfiguration;

public class ActiveItemsRowView extends HorizontalGridView implements OnHierarchyChangeListener {
    private final boolean mCardElevationSupported;
    private int mCardSpacing;
    AdapterDataObserver mChangeObserver;
    protected ViewDimmer.DimState mDimState;
    private boolean mIsAdjustable;
    private int mNumRows;
    private int mRowHeight;

    public ActiveItemsRowView(Context context) {
        this(context, null, 0);
    }

    public ActiveItemsRowView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ActiveItemsRowView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        this.mChangeObserver = new AdapterDataObserver() {
            public void onChanged() {
                ActiveItemsRowView.this.adjustNumRows();
                Adapter adapter = ActiveItemsRowView.this.getAdapter();
                if ((adapter instanceof AppsAdapter) && ((AppsAdapter) adapter).takeItemsHaveBeenSorted()) {
                    ActiveItemsRowView.this.setSelectedPosition(0);
                }
            }

            public void onItemRangeInserted(int positionStart, int itemCount) {
                ActiveItemsRowView.this.adjustNumRows();
            }

            public void onItemRangeRemoved(int positionStart, int itemCount) {
                ActiveItemsRowView.this.adjustNumRows();
            }
        };
        setChildrenDrawingOrderEnabled(true);
        setAnimateChildLayout(true);
        this.mDimState = ViewDimmer.DimState.INACTIVE;
        this.mCardElevationSupported = LauncherConfiguration.getInstance().isCardElevationEnabled();
    }

    public boolean hasOverlappingRendering() {
        return hasFocus() && super.hasOverlappingRendering();
    }

    public void setAdapter(Adapter adapter) {
        if (getAdapter() != null) {
            getAdapter().unregisterAdapterDataObserver(this.mChangeObserver);
        }
        super.setAdapter(adapter);
        if (adapter != null) {
            adapter.registerAdapterDataObserver(this.mChangeObserver);
        }
    }

    public void onChildAttachedToWindow(View child) {
        super.onChildAttachedToWindow(child);
        child.setActivated(ViewDimmer.dimStateToActivated(this.mDimState));
        if (child instanceof DimmableItem) {
            ((DimmableItem) child).setDimState(this.mDimState, false);
        }
        if (this.mCardElevationSupported) {
            child.setZ((float) getResources().getDimensionPixelOffset(R.dimen.unselected_item_z));
        }
    }

    public int getNumRows() {
        return this.mNumRows;
    }

    public void setActivated(boolean activated) {
        if (ViewDimmer.dimStateToActivated(this.mDimState) != activated) {
            this.mDimState = ViewDimmer.activatedToDimState(activated);
            int count = getChildCount();
            for (int i = 0; i < count; i++) {
                View view = getChildAt(i);
                if (view instanceof DimmableItem) {
                    ((DimmableItem) view).setDimState(this.mDimState, false);
                }
            }
        }
        super.setActivated(activated);
    }

    public boolean isRowActive() {
        return ViewDimmer.dimStateToActivated(this.mDimState);
    }

    public void setIsNumRowsAdjustable(boolean isAdjustable) {
        this.mIsAdjustable = isAdjustable;
        setOnHierarchyChangeListener(isAdjustable ? this : null);
    }

    public void adjustNumRows(int numRows, int cardSpacing, int rowHeight) {
        // Log.w("+++ adjustNumRows INIT", "" + numRows + " " + cardSpacing + " " + rowHeight);
        if (this.mIsAdjustable && this.mNumRows != numRows) {
            this.mNumRows = numRows;
            this.mCardSpacing = cardSpacing;
            this.mRowHeight = rowHeight;
            post(new Runnable() {
                public void run() {
                    ActiveItemsRowView.this.getLayoutParams().height = ((ActiveItemsRowView.this.mNumRows * ActiveItemsRowView.this.mRowHeight) + ((ActiveItemsRowView.this.mNumRows - 1) * ActiveItemsRowView.this.mCardSpacing)) + (ActiveItemsRowView.this.getPaddingTop() + ActiveItemsRowView.this.getPaddingBottom());
                    // Log.w("+++ adjustNumRows height", "" + ActiveItemsRowView.this.getLayoutParams().height);
                    ActiveItemsRowView.this.setNumRows(ActiveItemsRowView.this.mNumRows);
                    // Log.w("+++ adjustNumRows setNumRows", "" + ActiveItemsRowView.this.mNumRows);
                    ActiveItemsRowView.this.setRowHeight(ActiveItemsRowView.this.mRowHeight);
                    // Log.w("+++ adjustNumRows setRowHeight", "" + ActiveItemsRowView.this.mRowHeight);
                }
            });
        }
    }

    private void adjustNumRows() {
        int integer;
        if (this.mNumRows > 0)
            integer = this.mNumRows;
        else {
            Resources res = getResources();
            Context ctx = getContext();
            if (getAdapter().getItemCount() > RowPreferences.getAppsMax(ctx)) {
                integer = res.getInteger(R.integer.max_num_banner_rows);
            } else {
                integer = res.getInteger(R.integer.min_num_banner_rows);
            }
        }
        adjustNumRows(integer, this.mCardSpacing, this.mRowHeight);
    }

    public void childHasTransientStateChanged(View child, boolean hasTransientState) {
    }

    public void onChildViewAdded(View parent, View child) {
        adjustNumRows();
    }

    public void onChildViewRemoved(View parent, View child) {
        adjustNumRows();
    }

    public void onScrollStateChanged(int state) {
        super.onScrollStateChanged(state);
        enableAnimations(this);
    }

    private void enableAnimations(View view) {
        if (view instanceof ParticipatesInScrollAnimation) {
            ((ParticipatesInScrollAnimation) view).setAnimationsEnabled(true);
        }
        if (view instanceof ViewGroup) {
            int count = ((ViewGroup) view).getChildCount();
            for (int i = 0; i < count; i++) {
                enableAnimations(((ViewGroup) view).getChildAt(i));
            }
        }
    }

    public View focusSearch(View focused, int direction) {
        if (direction == 17 || direction == 66) {
            enableAnimations(this);
        }
        return super.focusSearch(focused, direction);
    }

}

File: app/src/main/java/com/amazon/tv/leanbacklauncher/ActiveItemsRowView.kt
package com.amazon.tv.leanbacklauncher

import android.content.Context
import android.util.AttributeSet
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.view.ViewGroup.OnHierarchyChangeListener
import androidx.leanback.widget.HorizontalGridView
import com.amazon.tv.firetv.leanbacklauncher.apps.AppCategory
import com.amazon.tv.firetv.leanbacklauncher.apps.RowPreferences
import com.amazon.tv.firetv.leanbacklauncher.apps.RowPreferences.getAppsColumns
import com.amazon.tv.firetv.leanbacklauncher.apps.RowType
import com.amazon.tv.leanbacklauncher.animation.ParticipatesInScrollAnimation
import com.amazon.tv.leanbacklauncher.animation.ViewDimmer
import com.amazon.tv.leanbacklauncher.animation.ViewDimmer.DimState
import com.amazon.tv.leanbacklauncher.apps.AppsAdapter
import com.amazon.tv.leanbacklauncher.capabilities.LauncherConfiguration
import kotlin.math.abs

open class ActiveItemsRowView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0
) : HorizontalGridView(context, attrs, defStyle), OnHierarchyChangeListener {
    private val TAG =
        if (BuildConfig.DEBUG) ("*" + javaClass.simpleName).take(21) else javaClass.simpleName
    private val mCardElevationSupported: Boolean
    private var mCardSpacing = 0
    private var mChangeObserver: AdapterDataObserver
    protected var mDimState: DimState
    private var mIsAdjustable = false
    var aNumRows = 0
        private set
    private var mRowHeight = 0
    private var rowType: RowType? = null

    init {
        mChangeObserver = object : AdapterDataObserver() {
            override fun onChanged() {
                val adapter = this@ActiveItemsRowView.adapter
                if (adapter is AppsAdapter)
                    rowType = adapter.getRowType()
                //if (BuildConfig.DEBUG) Log.d(TAG, "onChanged() rowType: $rowType aNumRows: $aNumRows")
                //this@ActiveItemsRowView.adjustNumRows()
                this@ActiveItemsRowView.adjustNumRows(rowType)
                if (adapter is AppsAdapter && adapter.takeItemsHaveBeenSorted()) {
                    this@ActiveItemsRowView.selectedPosition = 0
                    //if (BuildConfig.DEBUG) Log.d(TAG, "onChanged() setSelectedPosition(0)")
                }
            }

            override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
                val adapter = this@ActiveItemsRowView.adapter
                if (adapter is AppsAdapter)
                    rowType = adapter.getRowType()
                //if (BuildConfig.DEBUG) Log.d(TAG, "rowType: $rowType onItemRangeInserted(positionStart $positionStart, itemCount $itemCount)")
                //this@ActiveItemsRowView.adjustNumRows()
                this@ActiveItemsRowView.adjustNumRows(rowType)
            }

            override fun onItemRangeRemoved(positionStart: Int, itemCount: Int) {
                val adapter = this@ActiveItemsRowView.adapter
                if (adapter is AppsAdapter)
                    rowType = adapter.getRowType()
                //if (BuildConfig.DEBUG) Log.d(TAG, "rowType: $rowType onItemRangeRemoved(positionStart $positionStart, itemCount $itemCount)")
                //this@ActiveItemsRowView.adjustNumRows()
                this@ActiveItemsRowView.adjustNumRows(rowType)
            }
        }
        isChildrenDrawingOrderEnabled = true
        setAnimateChildLayout(true)
        mDimState = DimState.INACTIVE
        mCardElevationSupported = LauncherConfiguration.getInstance().isCardElevationEnabled
    }

    override fun hasOverlappingRendering(): Boolean {
        return hasFocus() && super.hasOverlappingRendering()
    }

    override fun setAdapter(adapter: Adapter<*>?) {
        getAdapter()?.unregisterAdapterDataObserver(mChangeObserver)
        super.setAdapter(adapter)
        adapter?.registerAdapterDataObserver(mChangeObserver)
    }

    override fun onChildAttachedToWindow(child: View) {
        super.onChildAttachedToWindow(child)
        child.isActivated = ViewDimmer.dimStateToActivated(mDimState)
        if (child is DimmableItem) {
            (child as DimmableItem).setDimState(mDimState, false)
        }
        if (mCardElevationSupported) {
            child.z = resources.getDimensionPixelOffset(R.dimen.unselected_item_z).toFloat()
        }
    }

    override fun setActivated(activated: Boolean) {
        if (ViewDimmer.dimStateToActivated(mDimState) != activated) {
            mDimState = ViewDimmer.activatedToDimState(activated)
            val count = childCount
            for (i in 0 until count) {
                val view = getChildAt(i)
                if (view is DimmableItem) {
                    (view as DimmableItem).setDimState(mDimState, false)
                }
            }
        }
        super.setActivated(activated)
    }

    val isRowActive: Boolean
        get() = ViewDimmer.dimStateToActivated(mDimState)

    fun setIsNumRowsAdjustable(isAdjustable: Boolean) {
        mIsAdjustable = isAdjustable
        setOnHierarchyChangeListener(if (isAdjustable) this else null)
    }

    // called on HomeScreenAdapter initAppRow
    fun adjustNumRows(numRows: Int, cardSpacing: Int, rowHeight: Int) {
        if (mIsAdjustable && this.aNumRows != numRows) {
            //if (BuildConfig.DEBUG) Log.w(TAG, "adjustNumRows[$numRows] was $aNumRows")
            this.aNumRows = numRows
            mCardSpacing = cardSpacing
            mRowHeight = rowHeight
            post {
                val lp = this@ActiveItemsRowView.layoutParams
                lp.height =
                    this@ActiveItemsRowView.aNumRows * mRowHeight + (this@ActiveItemsRowView.aNumRows - 1) * mCardSpacing + (this@ActiveItemsRowView.paddingTop + this@ActiveItemsRowView.paddingBottom)
                //if (BuildConfig.DEBUG) Log.w(TAG, "new height: ${this@ActiveItemsRowView.layoutParams.height}")
                setNumRows(this@ActiveItemsRowView.aNumRows)
                //if (BuildConfig.DEBUG) Log.w(TAG, "call setNumRows($aNumRows)")
                this@ActiveItemsRowView.setRowHeight(mRowHeight)
                //if (BuildConfig.DEBUG) Log.w(TAG, "set RowHeight($mRowHeight)")
            }
        }
    }

    // called on onChanged / onItemRangeInserted / onItemRangeRemoved / onChildViewAdded / onChildViewRemoved
    private fun adjustNumRows(type: RowType?) {
        type?.let {
            // calculate number of rows based on maxApps:
            // always fill a least one full row of maxApps
            val curApps = adapter?.itemCount ?: 0
            val maxCols = getAppsColumns(context)
            var base = abs(curApps / maxCols)
            val lost = (maxCols * (base + 1)) - curApps
            if (lost < base + 1) base += 1
            val userMax: Int = when (type) {
                RowType.APPS -> RowPreferences.getRowMax(AppCategory.OTHER, context)
                RowType.VIDEO -> RowPreferences.getRowMax(AppCategory.VIDEO, context)
                RowType.MUSIC -> RowPreferences.getRowMax(AppCategory.MUSIC, context)
                RowType.GAMES -> RowPreferences.getRowMax(AppCategory.GAME, context)
                RowType.FAVORITES -> RowPreferences.getFavoriteRowMax(context)
                else -> context.resources.getInteger(R.integer.max_num_banner_rows)
            }
            //if (BuildConfig.DEBUG) Log.d(TAG, "adjustNumRows($type) curApps:$curApps maxCols:$maxCols userMax:$userMax")
            // numRows
            val maxRows = if (base > 0) base.coerceAtMost(userMax) else resources.getInteger(R.integer.min_num_banner_rows)
            // apply numRows
            adjustNumRows(maxRows, mCardSpacing, mRowHeight)
        }
    }

//    private fun adjustNumRows() { // UNUSED
//        val integer: Int = if (aNumRows > 0) aNumRows else {
//            if (adapter!!.itemCount > getAppsColumns(context)) {
//                resources.getInteger(R.integer.max_num_banner_rows)
//            } else {
//                resources.getInteger(R.integer.min_num_banner_rows)
//            }
//        }
//        // apply
//        adjustNumRows(integer, mCardSpacing, mRowHeight)
//    }

    override fun childHasTransientStateChanged(child: View, hasTransientState: Boolean) {}

    override fun onChildViewAdded(parent: View, child: View) {
        //if (BuildConfig.DEBUG) Log.d(TAG, "onChildViewAdded() adjustNumRows($rowType)")
        //adjustNumRows()
        adjustNumRows(rowType)
    }

    override fun onChildViewRemoved(parent: View, child: View) {
        //if (BuildConfig.DEBUG) Log.d(TAG, "onChildViewRemoved() adjustNumRows($rowType)")
        //adjustNumRows()
        adjustNumRows(rowType)
    }

    override fun onScrollStateChanged(state: Int) {
        super.onScrollStateChanged(state)
        enableAnimations(this)
    }

    private fun enableAnimations(view: View) {
        if (view is ParticipatesInScrollAnimation) {
            (view as ParticipatesInScrollAnimation).setAnimationsEnabled(true)
        }
        if (view is ViewGroup) {
            val count = view.childCount
            for (i in 0 until count) {
                enableAnimations(view.getChildAt(i))
            }
        }
    }

    override fun focusSearch(focused: View, direction: Int): View {
        if (direction == FOCUS_LEFT || direction == FOCUS_RIGHT) {
            enableAnimations(this)
        }
        return super.focusSearch(focused, direction)
    }

}